### PR TITLE
Analytics: Refactor user identification for server side analytics

### PR DESCRIPTION
**Pull request by yurynix**
Tracks expect `_ut` and `_ui` set to record the event.

In certain scenarios, on server-side, those were missing.
This PR refactors the user identification to work as follows:

1. If a user has an auth cookie set, we get user identification for tracks from it.
else
2. if the request has query params `_ut` and `_ui` set, we get user identification for tracks from it.
3. If we don't have any of the above, we set `_ut` to 'anon' and `_ui` to a UUID v4.

 
#### Testing instructions
  
1. Run `git checkout update/csp-tracks` and start your server, or open a [live branch](https://calypso.live/?branch=update/csp-tracks)
2. Open the [`Login` page](http://calypso.localhost:3000/log-in)
3. Make sure a CSP report is triggered ( a POST to `/cspreport` )
4. Assert a tracks event was indeed recorded with the correct user
5. Assert a tracks event `calypso_stats_blocked` is recorded for a request to `calypso.localhost:3000/nostats.js?_ut=wpcom:user_id&_ui=12345` for a user type `wpcom:user_id` id `<your user id>` ( make sure you don't have the auth cookie set for this one )
  
#### Reviews
  
- [ ] Code
- [ ] Product

### CODE REVIEW

File: server/lib/analytics/index.js [hunk 1]
```diff
@@ -4,15 +4,46 @@
...
+			};
+		} catch ( ex ) {
+			// ignore the error and let it fallback to anonymous below
+		}
+	}
+
+	// if user doesn't have a cookie, and we had the user identification passed to us on query params
+	// we'll use that, otherwise, we'll just use anonymous.
+	const userType = get( request, 'query._ut', 'anon' );
+	const userId = get( request, 'query._ui', uuid() );
```
[the remark this answers is not in the document]
Comment (yurynix) [reply server/lib/analytics/index.js]: You're right, I'll refactor it 👍  Thanks!

File: server/lib/analytics/index.js [hunk 2]
```diff
@@ -4,15 +4,52 @@
...
+				_ut: 'wpcom:user_id',
+			};
+		} catch ( ex ) {
+			// ignore the error and let it fallback to anonymous below
+		}
+	}
+
+	// if user doesn't have a cookie, and we had the user identification passed to us on query params
+	// we'll use that, otherwise, we'll just use anonymous.
+	let userIdType = get( request, 'query._ut', 'anon' );
```
Comment (Tug) [inline server/lib/analytics/index.js]: does it matter that `anon` is set? Can't we leave it undefined?

Comment (yurynix) [reply server/lib/analytics/index.js]: Not sure what you meant exactly, userIdType ( `_ut` ) is a required parameter we need to pass to tracks

**Comment by Tug** [comment]
LGTM 👍

File: server/lib/analytics/index.js (hunk 2 above)
Comment (bperson) [reply server/lib/analytics/index.js]: If `_ut` is not set to one of the whitelisted values, the event will be rejected.

File: server/lib/analytics/index.js [hunk 3]
```diff
@@ -4,15 +4,52 @@
...
+	let userId = get( request, 'query._ui', null );
+
+	// if user ID not set, even though userIdType is set, we revert to anonymous
+	if ( ! userId ) {
```
Comment (bperson) [inline server/lib/analytics/index.js]: I think we'd be even safer with something like this:

```
// If we have a full identity, use it
if ( has( request, 'query._ut' ) && has( request, 'query._ui' ) ) {
    return {
        _ui: request.query._ui,
        _ut: request.query._ut
    };
}

// We didn't get a full identity, create an anon ID
return {
    _ut: 'anon',
    _ui: uuid(),
};
```

Though I hope we won't ever send half of the parameters used for a "full" identity, I'd rather we create a new anonymous ID than used `_ui` if `_ut` isn't set: either we have a full proper identity in `_ui` and `_ut`, or we discard it and create a proper "anonymous" identity?

**Comment by bperson** [comment]
I added another nitpick to handle half-broken identities,  but I don't think it's a blocker ( and even more so since the "old" code suffers from the same issue if it receives half the identity parameters ;) )

LGTM

File: server/lib/analytics/index.js (hunk 3 above)
Comment (yurynix) [reply server/lib/analytics/index.js]: Yep, you're right again, I've updated the code.

**Comment by yurynix** [comment]
Thank you for the review @bperson!

**Pull request merged**